From c15f3ef7c0df7300626790afc83d2f433b672e0a Mon Sep 17 00:00:00 2001
From: Nhan Nguyen <nhan.nguyen.yb@renesas.com>
Date: Mon, 13 Aug 2018 18:32:33 +0700
Subject: [PATCH 598/628] media: soc_camera: rcar_vin: Fix alignment of
 clipping size

    Since the Start Line Pre-Clip register, the Start Pixel Pre-Clip
    register and the End Line Post-Clip register do not have restriction
    of H/W to a setting value, the processing of alignment is unnecessary.
    This patch corrects so that processing of alignment is not performed.

    However, the End Pixel Post-Clip register has restriction
    of H/W which must be an even number value at the time of the
    output of YCbCr-422 format. By this patch, the processing of
    alignment to an even number value is added on the above-mentioned
    conditions.

    The variable set to a register is as follows.

     - Start Line Pre-Clip register
       cam->vin_top
     - Start Pixel Pre-Clip register
       cam->vin_left
     - End Line Post-Clip register
       pix->height
     - End Pixel Post-Clip register
       pix->width
    This commit is port from commit 90d322c24f95 ("media: soc_camera: rcar_vin: Fix alignment of clipping")

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Signed-off-by: Nhan Nguyen <nhan.nguyen.yb@renesas.com>
---
 drivers/media/platform/soc_camera/rcar_vin.c | 17 ++++++++++++-----
 1 file changed, 12 insertions(+), 5 deletions(-)

diff --git a/drivers/media/platform/soc_camera/rcar_vin.c b/drivers/media/platform/soc_camera/rcar_vin.c
index 75f9c43..0481f3c 100644
--- a/drivers/media/platform/soc_camera/rcar_vin.c
+++ b/drivers/media/platform/soc_camera/rcar_vin.c
@@ -1857,8 +1857,8 @@ static int rcar_vin_set_crop(struct soc_camera_device *icd,
 	cam->width = mf->width;
 	cam->height = mf->height;
 
-	cam->vin_left = rect->left & ~1;
-	cam->vin_top = rect->top & ~1;
+	cam->vin_left = rect->left;
+	cam->vin_top = rect->top;
 
 	/* Use VIN cropping to crop to the new window. */
 	ret = rcar_vin_set_rect(icd);
@@ -2079,9 +2079,16 @@ static int rcar_vin_try_fmt(struct soc_camera_device *icd,
 		pix->colorspace = icd->colorspace;
 	}
 
-	/* FIXME: calculate using depth and bus width */
-	v4l_bound_align_image(&pix->width, 2, VIN_MAX_WIDTH, 1,
-			      &pix->height, 4, VIN_MAX_HEIGHT, 2, 0);
+	/* When performing a YCbCr-422 format output, even if it performs
+	*odd number clipping by pixel post clip processing, it is outputted
+	*to a memory per even pixels.*/
+	if ((pixfmt == V4L2_PIX_FMT_NV16) || (pixfmt == V4L2_PIX_FMT_YUYV) ||
+		(pixfmt == V4L2_PIX_FMT_UYVY))
+		v4l_bound_align_image(&pix->width, 5, VIN_MAX_WIDTH, 1,
+					&pix->height, 2, VIN_MAX_HEIGHT, 0, 0);
+	else
+		v4l_bound_align_image(&pix->width, 5, VIN_MAX_WIDTH, 0,
+					&pix->height, 2, VIN_MAX_HEIGHT, 0, 0);
 
 	width = pix->width;
 	height = pix->height;
-- 
2.7.4

